File: project7_Lsg.py
## Project7
# Numpy/Pandas 와 R 비교

# 환경설정
import numpy

all = [var for var in globals() if var[0] != "_"]
for var in all:
    del globals()[var]

import numpy as np
import pandas as pd

# 공통 데이터
data1 = [1.5, 4.0, 2.0, 3.0]
data1
data2 = [[1, 2, 3, 4],[5, 6, 7, 8]]
data2

# 1. empty(numpy)
data1 = np.empty((2, 2)) # 2*2 배열함수 생성 함수
data1
np.empty((2, 4, 5))

# empty(pandas)
df_empty = pd.DataFrame({'A' : []}) # 인덱스가 비어있는것을 확인하는 함수
df_empty
df_empty.empty

df = pd.DataFrame({'A' : [np.nan]}) # 인덱스에 nan만 있으면 비어있는것으로 간주 하지 않는다.
df
df.empty
df.dropna().empty # NA를 삭제하면 인덱스가 비워진 상태로 출력

# 2. sum(numpy)
np.sum(data1) # 데이터 안에 들어있는 모든 원소의 합 추출 함수
np.sum(data2)
# sum(pandas)
df = pd.DataFrame([[1.4, np.nan], [7.1, -4.5],
              [np.nan, np.nan], [0.75, -1.3]],
                index=['a', 'b', 'c', 'd'],
                columns=['one', 'two'])
df

df.sum()
df.sum('index')
df.sum('columns') # nan = 0으로 자동변환

# 3. mean(numpy)
np.mean(data1) # 데이터 안에 있는 모든 원소의 평균
np.mean(data2)
# mean(pandas)
df
df.mean()
df.mean('index')
df.mean('columns')

# 4. zeros(numpy)
np.zeros(10) # 주어진 길이나 모양에 0이라는 값으로 배열 생성
np.zeros((3, 6))
np.zeros((2, 3, 2))
# zeros(pandas)
a = pd.DataFrame(np.zeros(10))
b = pd.DataFrame(np.zeros((3, 6)))
c = pd.DataFrame(np.zeros(2,3,2)) # 3차원 부터는 출력의 지원이 안된다.
print(a)
print(b)

# 5. var(numpy)
np.var(data1, ddof=1) # 데이터 안에 있는 원소들의 분산 값 / ddof=1 (표준편차를 계산할 때, n-1로 나누라는 의미)
np.var(data2, ddof=1) # https://www.abbreviationfinder.org/ko/acronyms/ddof.html#aim
# var(pandas)
df.var()
df.var('index')
df.var('columns')

# 6. min(numpy)
np.min(data1)
np.min(data2)
# min(pandas)
df.min()
df.min('index')
df.min('columns')

# 7. max(numpy)
np.max(data1)
np.max(data2)
# max(pandas)
df.max()
df.max('index')
df.max('columns')

# 8. cumsum(numpy)
np.cumsum(data1) # 데이터 안에 있는 원소들의 누적 합의 추출 함수 (sum과는 별개)
np.cumsum(data2)
# cumsum(pandas)
df
df.cumsum('index')
df.cumsum('columns')

# 9. random(numpy)
data = np.random.randn(2, 3) # 난수 생성
data

# random(pandas)
df1 = pd.DataFrame(np.random.randn(2, 3))
df1

# 10. sqrt(numpy)
np.sqrt(data1)
np.sqrt(data2)
# sqrt(pandas)
df.dropna(axis=0)
df.dropna(axis=1)
df.transform('sqrt')

File: project7_Lsg.R
# 비교 데이터 생성

# numpy 비교 데이터
data1 <- c(1.5, 4, 2, 3) 
data1

vec1 <- c(1:8)
data2 <- array(vec1, c(1, 4, 2))
data2


# pandas dataframe 비교 데이터
own <- c(1.4, 7.1, NA, 0.75)
two <- c(NA, -4.5, NA, -1.3)
df <- data.frame(own, two)
df
rownames(df) <- c('a', 'b', 'c', 'd')
df

# empty
install.packages('plyr') # empty함수를 쓰기 위한 패키지 설치
library(plyr)

# pandas 비교
empty(df)

# numpy 비교
empty(data1)
empty(data2) # 데이터 프레임이 비었는지 확인 하는 함수
# pandas의 empty의 함수와 동일하게 쓰인다.

# sum
# numpy 비교
sum(data1) # 데이터 안에 있는 모든 원소의 합
sum(data2)

# pandas 비교
sum(df$own, na.rm = T)
sum(df$two, na.rm = T)
rowSums(df, na.rm = T)
# numpy, pandas와 동일하게 쓰인다.

# mean
# numpy 비교
mean(data1) # 데이터 안에 있는 모든 원소의 평균
mean(data2)

# pandas 비교
mean(df$own, na.rm = T)
mean(df$two, na.rm = T)
rowMeans(df, na.rm = T)
# numpy, pandas와 동일하게 쓰인다.

# zeros
install.packages('phonTools')
library(phonTools)

zeros(10)
zeros(3, 6)
zeros(2, 3, 2) # 3차원 지원 출력이 안됨

# var
# numpy 비교
var(data1)
var(data2)

# pandas 비교
var(df$own, na.rm = T)
var(df$two, na.rm = T)
apply(df, 1, var, na.rm = TRUE)

# min
# numpy 비교
min(data1) # 데이터 안에 있는 원소중 제일 작은 값
min(data2)

# pandas 비교
min(df$own, na.rm = T)
min(df$two, na.rm = T)
apply(df, 1, min, na.rm = TRUE)

# max
# numpy 비교
max(data1) # 데이터 안에 있는 원소중 제일 큰 값
max(data2)

# pandas 비교
max(df$own, na.rm = T)
max(df$two, na.rm = T)
apply(df, 1, max, na.rm = TRUE)

# cumsum
# numpy 비교
cumsum(data1) # 데이터 안에 있는 원소들의 누적 합
cumsum(data2)

# pandas 비교 na값을 전처리 해야함 그래야 비교가능
is.na(df)
df1 <- na.omit(df)
cumsum(df1$own)
cumsum(df1$two)
apply(df1, 1, cumsum)
# numpy, pandas와 동일하게 쓰인다.

# random
rnorm(n=3, mean=0, sd=2)

# sqrt
# numpy 비교
sqrt(data1)
sqrt(data2)

# pandas 비교
sqrt(df$own)
sqrt(df$two)
apply(df, 1, sqrt)
